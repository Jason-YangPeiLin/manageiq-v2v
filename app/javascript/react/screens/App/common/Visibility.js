import React from 'react';
import PropTypes from 'prop-types';
import { noop } from 'patternfly-react';

// This component has no role on the page, and will be hidden from screen readers (role="presentation").
// Renders the children wrapped in visibility: hidden if the condition is met.
// Either way, the children take up space, visibility will not change the flow of the layout.
// The onKeyUp handler and role are used to satisfy eslint-plugin-jsx-a11y.
const Visibility = ({ children, hidden }) => (
  <span style={{ visibility: hidden ? 'hidden' : '' }} onKeyUp={noop} role="presentation">
    {children}
  </span>
);

Visibility.propTypes = {
  children: PropTypes.node,
  hidden: PropTypes.bool
};

export default Visibility;
