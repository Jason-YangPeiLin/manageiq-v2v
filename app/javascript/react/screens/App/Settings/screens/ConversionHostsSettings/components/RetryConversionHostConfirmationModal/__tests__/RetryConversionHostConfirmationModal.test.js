import React from 'react';
import { shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';
import { reducer as formReducer } from 'redux-form';

import { generateStore } from '../../../../../../common/testReduxHelpers';
import RetryConversionHostConfirmationModalConnected, {
  RetryConversionHostConfirmationModal
} from '../RetryConversionHostConfirmationModal';

const flushPromises = () => new Promise(setImmediate); // https://stackoverflow.com/a/51045733

const mockVddkTask = {
  name: 'Mock Host',
  context_data: { request_params: { mock: 'params', vmware_vddk_package_url: 'mock vddk path' } }
};
const mockSshTask = { name: 'Mock Host', context_data: { request_params: { mock: 'params' } } };

const mockVddkForm = {
  values: { conversionHostSshKey: { body: 'mock conversion host key' } }
};
const mockSshForm = {
  values: {
    conversionHostSshKey: { body: 'mock conversion host key' },
    vmwareSshKey: { body: 'mock vmware key' }
  }
};

describe('retry conversion host confirmation modal', () => {
  const store = generateStore({ form: formReducer }, {});

  const ProviderWrapper = ({ children }) => <Provider store={store}>{children}</Provider>; // eslint-disable-line react/prop-types

  const getBaseProps = () => ({
    show: true,
    conversionHostTaskToRetry: mockVddkTask,
    isPostingConversionHosts: false,
    hideConversionHostRetryModalAction: jest.fn(),
    conversionHostRetryModalExitedAction: jest.fn(),
    postConversionHostsAction: jest.fn(() => Promise.resolve()),
    postConversionHostsUrl: '/api/mock/conversion_hosts',
    retryForm: mockVddkForm,
    store
  });

  it('renders the redux-form wrapper correctly', () => {
    const component = shallow(<RetryConversionHostConfirmationModalConnected {...getBaseProps()} />);
    expect(component).toMatchSnapshot();
  });

  it('renders correctly for a VDDK task', () => {
    const component = shallow(<RetryConversionHostConfirmationModal {...getBaseProps()} />);
    expect(component.find('TextFileField')).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });

  it('renders correctly for an SSH task', () => {
    const component = shallow(
      <RetryConversionHostConfirmationModal
        {...getBaseProps()}
        conversionHostTaskToRetry={mockSshTask}
        retryForm={mockSshForm}
      />
    );
    expect(component.find('TextFileField')).toHaveLength(2);
    expect(component).toMatchSnapshot();
  });

  it('disables the retry button when there are validation errors', () => {
    const component = mount(
      <ProviderWrapper>
        <RetryConversionHostConfirmationModalConnected
          {...getBaseProps()}
          retryForm={{ syncErrors: { mock: 'errors' } }}
        />
      </ProviderWrapper>
    );
    expect(component.find('Button[bsStyle="primary"]').props().disabled).toBeTruthy();
  });

  it('disables the retry button when already retrying', () => {
    const component = mount(
      <ProviderWrapper>
        <RetryConversionHostConfirmationModalConnected {...getBaseProps()} isPostingConversionHosts />
      </ProviderWrapper>
    );
    expect(component.find('Button[bsStyle="primary"]').props().disabled).toBeTruthy();
  });

  it('constructs the postBody correctly and calls postConversionHostsAction for a VDDK task', async () => {
    const props = getBaseProps();
    const component = mount(
      <ProviderWrapper>
        <RetryConversionHostConfirmationModalConnected {...props} />
      </ProviderWrapper>
    );
    component.find('Button[bsStyle="primary"]').simulate('click');
    expect(props.postConversionHostsAction).toHaveBeenCalledTimes(1);
    expect(props.postConversionHostsAction).toHaveBeenCalledWith(props.postConversionHostsUrl, [
      {
        mock: 'params',
        vmware_vddk_package_url: 'mock vddk path',
        conversion_host_ssh_private_key: 'mock conversion host key'
      }
    ]);
    expect(props.hideConversionHostRetryModalAction).toHaveBeenCalledTimes(0);
    await flushPromises();
    expect(props.hideConversionHostRetryModalAction).toHaveBeenCalledTimes(1);
  });

  it('constructs the postBody correctly and calls postConversionHostsAction for an SSH task', async () => {
    const props = getBaseProps();
    const component = mount(
      <ProviderWrapper>
        <RetryConversionHostConfirmationModalConnected
          {...props}
          conversionHostTaskToRetry={mockSshTask}
          retryForm={mockSshForm}
        />
      </ProviderWrapper>
    );
    component.find('Button[bsStyle="primary"]').simulate('click');
    expect(props.postConversionHostsAction).toHaveBeenCalledTimes(1);
    expect(props.postConversionHostsAction).toHaveBeenCalledWith(props.postConversionHostsUrl, [
      {
        mock: 'params',
        conversion_host_ssh_private_key: 'mock conversion host key',
        vmware_ssh_private_key: 'mock vmware key'
      }
    ]);
    expect(props.hideConversionHostRetryModalAction).toHaveBeenCalledTimes(0);
    await flushPromises();
    expect(props.hideConversionHostRetryModalAction).toHaveBeenCalledTimes(1);
  });
});
