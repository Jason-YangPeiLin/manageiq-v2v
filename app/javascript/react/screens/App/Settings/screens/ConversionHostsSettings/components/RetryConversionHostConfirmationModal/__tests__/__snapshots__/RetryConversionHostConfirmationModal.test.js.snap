// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`retry conversion host confirmation modal renders correctly for a VDDK task 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop="static"
  bsClass="modal"
  dialogClassName="modal-lg wizard-pf"
  dialogComponentClass={[Function]}
  enforceFocus={true}
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onExited={[MockFunction]}
  onHide={[MockFunction]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    bsClass="modal-header"
    closeButton={false}
    closeLabel="Close"
  >
    <ModalCloseButton
      className=""
      closeText="Close"
      onClick={[MockFunction]}
    />
    <ModalTitle
      bsClass="modal-title"
      componentClass="h4"
    >
      Retry Conversion Host Configuration
    </ModalTitle>
  </ModalHeader>
  <ModalBody
    bsClass="modal-body"
    componentClass="div"
  >
    <Form
      bsClass="form"
      className="form-horizontal"
      componentClass="form"
      horizontal={false}
      inline={false}
    >
      <ul>
        <p>
          For your security, SSH private keys from the configuration wizard were not saved. Re-enter the necessary authentication details to retry configuration of the following conversion host:
        </p>
      </ul>
      <FormGroup
        bsClass="form-group"
      >
        <Col
          bsClass="col"
          componentClass={[Function]}
          sm={4}
        >
          Host
        </Col>
        <Col
          bsClass="col"
          componentClass="div"
          sm={7}
          style={
            Object {
              "paddingTop": 3,
            }
          }
        >
          Mock Host
        </Col>
      </FormGroup>
      <TextFileField
        controlId="host-ssh-key-input"
        controlWidth={7}
        help="Upload your SSH key file or paste its contents below."
        hideBody={false}
        info=""
        label="Conversion Host SSH private key"
        labelWidth={4}
        name="conversionHostSshKey"
      />
    </Form>
  </ModalBody>
  <ModalFooter
    bsClass="modal-footer"
    componentClass="div"
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="default"
      className="btn-cancel"
      disabled={false}
      onClick={[MockFunction]}
    >
      Cancel
    </Button>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="primary"
      disabled={false}
      onClick={[Function]}
    >
      Retry
    </Button>
  </ModalFooter>
</Modal>
`;

exports[`retry conversion host confirmation modal renders correctly for an SSH task 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop="static"
  bsClass="modal"
  dialogClassName="modal-lg wizard-pf"
  dialogComponentClass={[Function]}
  enforceFocus={true}
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onExited={[MockFunction]}
  onHide={[MockFunction]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    bsClass="modal-header"
    closeButton={false}
    closeLabel="Close"
  >
    <ModalCloseButton
      className=""
      closeText="Close"
      onClick={[MockFunction]}
    />
    <ModalTitle
      bsClass="modal-title"
      componentClass="h4"
    >
      Retry Conversion Host Configuration
    </ModalTitle>
  </ModalHeader>
  <ModalBody
    bsClass="modal-body"
    componentClass="div"
  >
    <Form
      bsClass="form"
      className="form-horizontal"
      componentClass="form"
      horizontal={false}
      inline={false}
    >
      <ul>
        <p>
          For your security, SSH private keys from the configuration wizard were not saved. Re-enter the necessary authentication details to retry configuration of the following conversion host:
        </p>
      </ul>
      <FormGroup
        bsClass="form-group"
      >
        <Col
          bsClass="col"
          componentClass={[Function]}
          sm={4}
        >
          Host
        </Col>
        <Col
          bsClass="col"
          componentClass="div"
          sm={7}
          style={
            Object {
              "paddingTop": 3,
            }
          }
        >
          Mock Host
        </Col>
      </FormGroup>
      <TextFileField
        controlId="host-ssh-key-input"
        controlWidth={7}
        help="Upload your SSH key file or paste its contents below."
        hideBody={false}
        info=""
        label="Conversion Host SSH private key"
        labelWidth={4}
        name="conversionHostSshKey"
      />
      <TextFileField
        controlId="vmware-ssh-key-input"
        controlWidth={7}
        help="Upload your SSH key file or paste its contents below."
        hideBody={false}
        label="VMware hypervisors SSH private key"
        labelWidth={4}
        name="vmwareSshKey"
        style={
          Object {
            "marginTop": 25,
          }
        }
      />
    </Form>
  </ModalBody>
  <ModalFooter
    bsClass="modal-footer"
    componentClass="div"
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="default"
      className="btn-cancel"
      disabled={false}
      onClick={[MockFunction]}
    >
      Cancel
    </Button>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="primary"
      disabled={false}
      onClick={[Function]}
    >
      Retry
    </Button>
  </ModalFooter>
</Modal>
`;

exports[`retry conversion host confirmation modal renders the redux-form wrapper correctly 1`] = `
<Connect(Form(RetryConversionHostConfirmationModal))
  conversionHostRetryModalExitedAction={[MockFunction]}
  conversionHostTaskToRetry={
    Object {
      "context_data": Object {
        "request_params": Object {
          "mock": "params",
          "vmware_vddk_package_url": "mock vddk path",
        },
      },
      "name": "Mock Host",
    }
  }
  destroyOnUnmount={true}
  enableReinitialize={false}
  forceUnregisterOnUnmount={false}
  form="retryConversionHost"
  getFormState={[Function]}
  hideConversionHostRetryModalAction={[MockFunction]}
  initialValues={
    Object {
      "conversionHostSshKey": Object {
        "body": "",
        "filename": "",
      },
      "vmwareSshKey": Object {
        "body": "",
        "filename": "",
      },
    }
  }
  isPostingConversionHosts={false}
  keepDirtyOnReinitialize={false}
  persistentSubmitErrors={false}
  postConversionHostsAction={[MockFunction]}
  postConversionHostsUrl="/api/mock/conversion_hosts"
  pure={true}
  retryForm={
    Object {
      "values": Object {
        "conversionHostSshKey": Object {
          "body": "mock conversion host key",
        },
      },
    }
  }
  shouldAsyncValidate={[Function]}
  shouldError={[Function]}
  shouldValidate={[Function]}
  shouldWarn={[Function]}
  show={true}
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
  touchOnBlur={true}
  touchOnChange={false}
  updateUnregisteredFields={false}
/>
`;
